xapian-applications/omega/postprocess.py

```diff
@@ -0,0 +1,187 @@
+            if row[QUERY] in queries:
+                continue
+
+            queries.add(row[QUERY])
+
+            # Convert Hitlist from str to list
+            hits = row[HITS]
+            hits = hits.strip().split(',')
+            row[HITS] = hits
+
+            clicklist = hits[:]
+
+            # Update clicklist with click values stored in map.
+            if row[QUERYID] in qid_to_clicks:
+                did_to_count = qid_to_clicks[row[QUERYID]]
+                for index, did in enumerate(clicklist):
+                    if did in did_to_count:
+                        clicklist[index] = '%s:%i' % (did, did_to_count[did])
+                    else:
+                        clicklist[index] = did + ':0'
+            else:
+                for index, did in enumerate(clicklist):
+                    if clicklist[index] == '':
+                        continue
+                    clicklist[index] = did + ':0'
+
+            # Serialise "Hits" and "Clicks"
+            row[HITS] = ','.join(row[HITS])
+            clicklist = ','.join(clicklist)
+
+            row.append(clicklist)
+            writer.writerow(row)
+
+
+def generate_query_file(final_log, query_file):
+    """Generates Query file formatted as per Xapian Letor documentation.
+
+    Input Args:
+        final_log (string): Path to final.log file.
+        query_file (string): Path to query.txt file.
+
+    Example (tab-delimited) entries in final.log:
+
+    QueryID,Query,Hits,Offset,Clicks
+    821f03288846297c2cf43c34766a38f7,book,"45,36,14,54,42,52,2,3,15,32",0,"45:0,36:0,14:0,54:2,42:0,52:0,2:0,3:0,15:0,32:0"
+    098f6bcd4621d373cade4e832627b4f6,test,"35,47,31,14,45,19,50,43,30,44",0,"35:1,47:0,31:0,14:0,45:0,19:0,50:0,43:0,30:0,44:0"
+
+    Example (comma-delimited) entries in query.txt:
+
+    821f03288846297c2cf43c34766a38f7,book
+    098f6bcd4621d373cade4e832627b4f6,test
+    """
+    with open(final_log, 'r') as s, open(query_file, 'w+') as w:
+        reader = csv.DictReader(s)
+        writer = csv.writer(w)
+
+        for row in reader:
+            writer.writerow([row['QueryID'], row['Query']])
+
+
+def test():
+    pass
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description='''Postprocess click data files.
+
+        This script generates the final clickstream log file from input search and
+        click log files and creates Query file that can be used by Xapian Letor
+        module for generating its training files.
+        ''', formatter_class=argparse.RawTextHelpFormatter)
+    parser.add_argument("search_log", type=str, help="Path to the search.log file.")
+    parser.add_argument("clicks_log", type=str, help="Path to the clicks.log file.")
+    parser.add_argument("final_log", type=str, help="Path to save final.log file.")
+    parser.add_argument("query_file", type=str, help="Path to save query.txt file.")
+    parser.add_argument("--test", type=test, help="Run tests for this script.")
```

Comment:
@jaylett Is this the right way for implementing `--test` to run test function defined in line 159? I'll go ahead and write actual tests in that function if this approach seems fine.